Software/src/PulseAudioSoundManager.cpp

```diff
@@ -0,0 +1,485 @@
+				const float c2 = cos((4*M_PI*i)/(samples - 1));
+				const float c3 = cos((6*M_PI*i)/(samples - 1));
+				dest[i] = 0.35875 - 0.48829*c1 + 0.14128*c2 - 0.001168*c3;
+			}
+			break;
+		case WINDOW_FLAT:
+			for (int i = 0; i < samples; i++)
+				dest[i] = 1.0f;
+			break;
+		case WINDOW_WELCH:
+			for (int i = 0; i < samples; i++)
+				dest[i] = 1 - pow((i - ((samples - 1)/2.0f))/((samples - 1)/2.0f), 2.0f);
+			break;
+		default:
+			for (int i = 0; i < samples; i++)
+				dest[i] = 0.0f;
+			break;
+	}
+	float sum = 0.0f;
+	for (int i = 0; i < samples; i++)
+		sum += dest[i];
+	for (int i = 0; i < samples; i++)
+		dest[i] /= sum;
+}
+
+static void pa_context_state_cb(pa_context *c, void *userdata)
+{
+	pa_context_state_t state;
+	int *pa_ready = (int *)userdata;
+
+	state = pa_context_get_state(c);
+	switch (state) {
+		case PA_CONTEXT_UNCONNECTED:
+			*pa_ready = 3;
+			break;
+		case PA_CONTEXT_CONNECTING:
+		case PA_CONTEXT_AUTHORIZING:
+		case PA_CONTEXT_SETTING_NAME:
+		default:
+			break;
+		case PA_CONTEXT_FAILED:
+		case PA_CONTEXT_TERMINATED:
+			*pa_ready = 2;
+			break;
+		case PA_CONTEXT_READY:
+			*pa_ready = 1;
+			break;
+	}
+}
+
+PulseAudioSoundManager::PulseAudioSoundManager(QObject *parent) : SoundManagerBase(parent)
+{
+	m_timer.setTimerType(Qt::PreciseTimer);
+	connect(&m_timer, SIGNAL(timeout()), this, SLOT(updateColors()));
+
+	m_pa_alive_timer.setTimerType(Qt::PreciseTimer);
+	connect(&m_pa_alive_timer, &QTimer::timeout, this, &PulseAudioSoundManager::checkPulse);
+}
+
+PulseAudioSoundManager::~PulseAudioSoundManager()
+{
+	m_cont = 0;
+	Uninit();
+}
+
+void PulseAudioSoundManager::Uninit()
+{
+	std::lock_guard<std::mutex> l(m_mutex);
+
+	if (m_stream) {
+		pa_stream_disconnect(m_stream);
+		pa_stream_unref(m_stream);
+		m_stream = nullptr;
+	}
+	if (m_context) {
+		pa_context_disconnect(m_context);
+		pa_context_unref(m_context);
+		m_context = nullptr;
+	}
+	if (m_main_loop) {
+		pa_threaded_mainloop_stop(m_main_loop);
+		pa_threaded_mainloop_free(m_main_loop);
+		m_main_loop = nullptr;
+	}
+
+	if (m_plan) {
+		fftw_destroy_plan(m_plan);
+		m_plan = nullptr;
+	}
+
+	if (m_output) {
+		fftw_free(m_output);
+		m_output = nullptr;
+	}
+
+	m_isInited = false;
+}
+
+bool PulseAudioSoundManager::init() {
+	int ret = 0;
+	m_buffer_samples = fftSize() * 2;
+	m_cont = 1;
+	m_ss.format = PA_SAMPLE_FLOAT32LE;
+	m_ss.rate = 44100;
+	m_ss.channels = 1;
+
+	init_buffers();
+
+	m_main_loop = pa_threaded_mainloop_new();
+	m_context = pa_context_new(pa_threaded_mainloop_get_api(m_main_loop), "Prismatik");
+	pa_context_set_state_callback(m_context, context_state_cb, this);
+
+	// Lock the mainloop so that it does not run and crash before the context is ready
+	pa_threaded_mainloop_lock(m_main_loop);
+	pa_threaded_mainloop_start(m_main_loop);
+
+	ret = pa_context_connect(m_context, m_server, PA_CONTEXT_NOFLAGS, NULL);
+
+	qInfo() << "pa_context_connect:" << pa_strerror(ret);
+	if (ret != PA_OK)
+		goto unlock_and_fail;
```

Comment:
any particular reason for using `goto` over a `throw`/`catch` combo?

Reply:
None, just going with C-isms.